### PR TITLE
port TwoPaneView to VB.NET

### PR DESCRIPTION
**Pull request by mrlacey**
# PR checklist

Quick summary of changes

This started out as a port of the TwoPaneView templates to Visual Basic. 
However, when testing the changes, the automated tests found numerous other issues. I started to also address these issues so that I could be confident that the tests were also validating everything I had added too.
There are still some tests that don't pass but they are not directly related to the templates added here. I have raised #3722 as a follow-up to address the other test issues.

- Which issue does this PR relate to? #3676 
- Anything that requires particular review or attention? Added fix for broken test infrastructure. Now fixed, have also patched other broken templates.
- Do all automated tests pass? the ones relating to changes do.
- Have automated tests been added for new features? n/a
- If you've changed the UI:
  - Be sure you are including screenshots to show the changes. n/a
  - Be sure you have reviewed the [accessibility checklist](accessibility.md). n/a
- If you've included a new template:
  - Be sure you reviewed the [Template Verification Checklist](https://github.com/microsoft/WindowsTemplateStudio/wiki/Checklist:-Template-Verification). yes
  - Be sure it's included in the list on this [UWP](https://github.com/microsoft/WindowsTemplateStudio/blob/dev/docs/UWP/getting-started-endusers.md) or [WPF](https://github.com/microsoft/WindowsTemplateStudio/blob/dev/docs/WPF/getting-started-endusers.md) getting started doc. - It is now (should have been added for C# templates)
- Have you raised issues for any needed follow-on work? n/a
- Have docs been updated? yes
- If breaking changes or different ways of doing things have been introduced, have they been communicated widely? n/a

### CODE REVIEW

**Comment by sibille** [comment]
ready for review?

**Comment by mrlacey** [comment]
> Ready for review?

yes

File: templates/Uwp/_comp/MVVMLight/Page.SplitView.TabbedNav.TwoPaneView._VB/Services/NavigationServiceEx_postaction.vb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        Private Sub UnregisterFrameEvents()
+            If _frame IsNot Nothing Then
+                RemoveHandler _frame.Navigated, AddressOf Frame_Navigated
+'{[{
+                RemoveHandler _frame.Navigating, AddressOf Frame_Navigating
+'}]}
+            End If
+        End Sub
+
+        Public Sub Frame_Navigated(sender As Object, e As NavigationEventArgs)
```
Comment (sibille) [inline templates/Uwp/_comp/MVVMLight/Page.SplitView.TabbedNav.TwoPaneView._VB/Services/NavigationServiceEx_postaction.vb:56]: This should be: 
```
        Public Sub Frame_Navigated(sender As Object, e As NavigationEventArgs)
'^^
'{[{
            Dim element As FrameworkElement = Nothing, backNavigationHandler As IBackNavigationHandler = Nothing

            element = TryCast(Frame.Content, FrameworkElement)
            If element IsNot Nothing Then
                backNavigationHandler = TryCast(element.DataContext, IBackNavigationHandler)
                If backNavigationHandler IsNot Nothing Then
                    AddHandler backNavigationHandler.OnPageCanGoBackChanged, AddressOf OnPageCanGoBackChanged
                End If
            End If
'}]}

            RaiseEvent Navigated(sender, e)
        End Sub
```

**Comment by mrlacey** [comment]
> I like the new icon :)

I assumed the master/detail one was a mistake and so asked the WinUI team what to use.

**Comment by mrlacey** [comment]
> Found an issue when adding the TwoPaneView twice 

 😢  
I spent days on other issues relating to adding things twice and getting all tests to pass.
I think a review of all testing as part of #3722 is in order.

**Comment by mrlacey** [comment]
This issue came about due to trying to copy [this](https://github.com/microsoft/WindowsTemplateStudio/blob/007f017fa0e0dc566c7512151d65e3dd59a8ad88/templates/Uwp/_comp/MVVMBasic/Page.SplitView.TabbedNav.TwoPaneView/Services/NavigationService_postaction.cs#L55) from the C# code that uses the removal of a one-line lambda function.
With no direct single-line equivalent in VB I ended up with something different that eventually became what was in the original PR due to issues with doing line removals for multiple lines at once.
There are no templates that currently remove more than one line at a time but it's possible (likely?) that there is something unexpected happening when trying to remove multiple lines at once. It's a potential rabbit-hole to explore now as it's not something that's needed (yet) but worth trying to remember if it ever is.

**Comment by sibille** [comment]
I first thought there's an issue in the postaction code, but then thinking about it, it works like excected, although you've hit an edge case for sure. 
First page removes a line that is re-added by the same code, but in a different position. Second page then deletes the line added by the first page, when comparing the code blocks it determines that it's not the same (as the last line is not there), that's why it gets inserted again.
What we can do is add a test to ensure that a postaction does not delete a line and insert it again, as this leads to code duplication.

File: templates/Uwp/_comp/CodeBehind/Page.SplitView.TabbedNav.TwoPaneView._VB/Services/NavigationService_postaction.vb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        Private Sub UnregisterFrameEvents()
+            If _frame IsNot Nothing Then
+                RemoveHandler _frame.Navigated, AddressOf Frame_Navigated
+'{[{
+                RemoveHandler _frame.Navigating, AddressOf Frame_Navigating
+'}]}
+            End If
+        End Sub
+
+        Public Sub Frame_Navigated(sender As Object, e As NavigationEventArgs)
```
[the remark this answers is not in the document]
Comment (mrlacey) [reply templates/Uwp/_comp/CodeBehind/Page.SplitView.TabbedNav.TwoPaneView._VB/Services/NavigationService_postaction.vb:49]: done

**Comment by mrlacey** [comment]
> What we can do is add a test to ensure that a postaction does not delete a line and insert it again, as this leads to code duplication.

This may be useful to help debugging issues but a better test to add will be to test that when including a page twice the code that is generated compiles. At the moment there are tests for one page compiling and two pages generating without issue.

**Pull request merged**